Restart units after the rpm transaction

keszybz opened the pull request:
/cc @Conan-Kudo

Conan-Kudo commented on src/rpm/triggers.systemd.in:
```diff
@@ -56,33 +44,31 @@ if posix.access("%{_localstatedir}/lib/rpm-state/systemd/needs-reload") then
--- specified users automatically. The priority is set such that it
--- will run before the tmpfiles file trigger.
+%transfiletriggerpostun -P 10000 -p <lua> -- @systemunitdir@ /etc/systemd/system
+-- We restart remaing services that should be restarted here.
```
`remaing` -> `remaining`

Conan-Kudo commented on src/rpm/triggers.systemd.in:
```diff
@@ -56,33 +44,31 @@ if posix.access("%{_localstatedir}/lib/rpm-state/systemd/needs-reload") then
--- This script will process files installed in @sysusersdir@ to create
--- specified users automatically. The priority is set such that it
--- will run before the tmpfiles file trigger.
+%transfiletriggerpostun -P 10000 -p <lua> -- @systemunitdir@ /etc/systemd/system
```
This should run _after_ all other systemd triggers, since we may need tmpfiles and users populated.

keszybz replied:
Hmm, is there some place where **all** the different trigger types are documented? I.e. how does %post, %postun, %triggerin, %transfiletriggerin, %triggerun, %transfiletriggerpostun, %postrans, %pretrans, and with different priorities, get sorted?

Conan-Kudo replied:
🤷🏾‍♂️ 

@pmatilai has been documenting things, maybe he's got it written up somewhere?

keszybz replied:
> This should run _after_ all other systemd triggers, since we may need tmpfiles and users populated.

10000 is super low, so it would be *after* most stuff, no?

bluca commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -2160,8 +2171,13 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
+      changed since the configuration was read and hence whether a configuration reload is recommended.
+      </para>
+
+      <para><varname>NeedsRestart</varname> is a boolean that can be externally to indicate that the service
```
"that can be set" ?

bluca commented on src/systemctl/systemctl.c:
```diff
@@ -38,6 +38,7 @@
 #include "systemctl-list-units.h"
 #include "systemctl-log-setting.h"
 #include "systemctl-logind.h"
+#include "systemctl-needs-restart.h"
```
nit: should be sorted one line below

poettering reviewed (changes requested):
hmm, if we add this, should we add a similar flag for "needs-reload"? i mean, it's not useful for the rpm case, but almost certainly for config mgmt systems?

i.e. i think we should add booleans for both cases, needs-reload, and needs-restart. and i'd probably make needs-daemons-reload a settable boolean too, because it's so similar, and it has should have similar usecases. 

I'd then call the new dbus call "FulfillNeeds" or so, and it issues the daemon-reload first, and then issues the reload and restart jobs...

a package manager would then call

```
systemctl set-property foo.service NeedDaemonReload=true NeedRestart=true
```

and a config manager would issue:

```
systemctl set-property foo.service NeedReload=true NeedDaemonReload=true
```

or so. you get the idea.

keszybz commented on src/core/unit.c:
```diff
@@ -3687,6 +3687,30 @@ static int unit_deserialize_job(Unit *u, FILE *f) {
+                };                                                      \
+                matched;                                                \
+        })
+
```
I think the approach with macros is OK here. The macros are defined right above the function where they are used, so they don't need to be super generic. Also, they will not be used outside of that function. And then there's the thing with bitmask fields in Unit: we can't really manage them through a function, at least not in any convenient way.

I now add new src/core/unit-serialize.[ch] files. The unit.c file was already rather long. And now the serialization stuff is a bit more hidden.

keszybz commented on src/core/dbus-unit.c:
```diff
@@ -864,6 +864,7 @@ const sd_bus_vtable bus_unit_vtable[] = {
         SD_BUS_PROPERTY("OnFailureJobMode", "s", property_get_job_mode, offsetof(Unit, on_failure_job_mode), SD_BUS_VTABLE_PROPERTY_CONST),
         SD_BUS_PROPERTY("IgnoreOnIsolate", "b", bus_property_get_bool, offsetof(Unit, ignore_on_isolate), SD_BUS_VTABLE_PROPERTY_CONST),
         SD_BUS_PROPERTY("NeedDaemonReload", "b", property_get_need_daemon_reload, 0, SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_WRITABLE_PROPERTY("NeedsRestart", "b", bus_property_get_bool, bus_property_set_bool, offsetof(Unit, needs_restart), 0),
```
So I shied away from SetUnitProperties / set-properties because those calls like to store the unit state on disk. But I'll just add a check to avoid that for NeedsRestart/NeedsReload. You are right, SetUnitProperties is a better interface for this.

keszybz commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -1685,6 +1685,9 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
       readonly b NeedDaemonReload = ...;
+      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
+      @org.freedesktop.systemd1.Privileged("true")
+      readwrite b NeedsRestart = ...;
```
I thought about this, but the the old naming is broken :( "NeedRestart" is just wrong. This is supposed to be an attribute so "this unit needs a restart", "this unit needs a daemon reload". Without the "s" it doesn't make sense gramatically.

keszybz replied:
Actually, "NeedDaemonReload" is okay-ish. Because even though it's a unit property, it applies to the whole manager, because we can't yet reload individual units. So "we need a daemon reload" can be abbreviated to "NeedDaemonReload", maybe. But the same doesn't hold for NeedsRestart.

keszybz commented on src/core/dbus-manager.c:
```diff
@@ -1807,6 +1807,63 @@ static int method_get_dynamic_users(sd_bus_message *message, void *userdata, sd_
+        if (r < 0)
+                return r;
+
+        return sd_bus_send(NULL, reply, NULL);
```
Supressing errors here doesn't seem nice either. In particular this can fail because of permission errors, and I'm pretty sure we should report those. I left it as is for now. But if you have a good idea how to change the code to ignore the right type of errors, please make a suggestion.

poettering replied:
well, there's also the other option: continue the loop, but then tell the caller that some enqueuing failed, and to check the logs for details. i.e. it's unlikely we can ever propagate all errors to the client via dbus error messages...

i.e. intrdouce a new dbus error: JobsPartiallyEnqueued() and in the log message say "Enqeueing some jobs failed, for details see logs".

Or maybe include the first error message that was encountered, but not all.

keszybz replied:
Done. I made it return the first error and a message:
"Failed to enqueue some jobs, see logs for details: %m".

keszybz commented:
Reworked to use set-property and also add Reload handling. I didn't do the thing with making NeedsDaemonReload settable because this shouldn't be set by user. We *know* which config we read, and we should include all of that config in the check for NeedsDaemonReload. Also, I hope that one day we can get rid of the global reload towards reloading config for individual units. So I don't want to add an interface to make this more prominent.

PTAL.

keszybz commented:
mkosi / ci (arch) looks unrelated:
```
exec-basic.service: Failed to attach to cgroup /app.slice/exec-basic.service: Operation not permitted
exec-basic.service: Failed at step CGROUP spawning touch: Operation not permitted
Received SIGCHLD from PID 8270 ((touch)).
Child 8270 ((touch)) died (code=exited, status=219/CGROUP)
exec-basic.service: Child 8270 belongs to exec-basic.service.
exec-basic.service: Main process exited, code=exited, status=219/CGROUP
exec-basic.service: Failed with result 'exit-code'.
exec-basic.service: Service will not restart (restart setting)
exec-basic.service: Changed start -> failed
exec-basic.service: Unit entered failed state.
Spawning thread to nuke /tmp/systemd-private-15b5bb8668e148ddbd1f3126f16d67ef-exec-basic.service-D828qL
Spawning thread to nuke /var/tmp/systemd-private-15b5bb8668e148ddbd1f3126f16d67ef-exec-basic.service-Psr01N
src/test/test-execute.c:722:test_exec_basic: exec-basic.service: exit status 219, expected 0
```

poettering commented:
maybe rename NeedDaemonReload →NeedsDaemonReload then? i.e. go the other way?

poettering reviewed (changes requested):
hmm, maybe this shouldn't be two bools, but a single enum: MarkFor with values "no", "reload" or "restart".

of these "restart" is the highest value, and once it's set it cannot be undone anymore (until processed at least), i.e. "reload" upgrades to "restart" but never reverts back to it.

and when processing them I figure we should enqeue the restarts first, and the reload seconds (why? because restarts might trigger other restarts, and thus make the reload unnecessary)

keszybz commented on src/core/unit-serialize.c:
```diff
@@ -172,6 +172,32 @@ static int unit_deserialize_job(Unit *u, FILE *f) {
+                if (_deserialize_matched) {                             \
+                        int _deserialize_r = parse_func(v, &target);    \
+                        if (_deserialize_r < 0)                         \
+                                log_unit_debug_errno(u, _deserialize_r, \
```
I'll leave this as is for now. Hopefully #11484 will be merged.

keszybz replied:
Actually, I removed the errno usage here. With the latest version of #11484, -EINVAL is always returned on error, so %m doesn't tell us anything useful.

keszybz commented on src/core/unit-serialize.c:
```diff
@@ -0,0 +1,810 @@
+        assert(mask == 0);
+}
+
+void unit_dump(Unit *u, FILE *f, const char *prefix) {
```
I think it's nice to keep unit_serialize and unit_dump in the same file. They are rather similar, and when on is touched, the other will usually be too. The whole file is 700 lines right now, which I think is OK.

keszybz commented:
Reworked. There's now a single StartAutomaticJobs() method that queues up restart or reload jobs. It can be invoked using `systemctl reload-or-restart --needing`. The two properties are merged into a single flags field (64 bits, though only two bits are used right now). I exposed the bits as an hexadecimal integer. The idea is that those bits are managed internally anyway, so the manager must understand those bits. They are are set using the same interface as before: `systemctl set-property xxx.service NeedsRestart=true NeedsReload=true`. I think this makes the interface nice to use in a script and such, while making it very easy to extend with additional flags in the future.

I also changed the method to continue and return the first error as discussed above. Most of the series is completely unchanged. In the rpm stuff, I only changed `restart --needing` to `reload-or-restart --needing`.

keszybz commented:
packit i386:
```
 # /usr/bin/dnf --installroot /var/lib/mock/fedora-rawhide-i686-bootstrap-1612981284.384537/root/ --releasever 34 --setopt=deltarpm=False --allowerasing --disableplugin=local --disableplugin=spacewalk install dnf dnf-plugins-core --setopt=tsflags=nocontexts --setopt=tsflags=nocontexts --setopt=tsflags=nocontexts
No matches found for the following disable plugin patterns: local, spacewalk
Copr repository                                 158 kB/s | 3.3 kB     00:00    
local                                            40 MB/s |  14 MB     00:00    
Errors during downloading metadata for repository 'local':
  - Status code: 404 for https://kojipkgs.fedoraproject.org/repos/rawhide/latest/i386/repodata/6f54197d96a5f5c4f5e72d0b1853a65b97367094ac43a39b7c4bd5dd891ba3a5-filelists.xml.gz (IP: 38.145.60.20)
Error: Failed to download metadata for repo 'local': Yum repo downloading error: Downloading error(s): repodata/6f54197d96a5f5c4f5e72d0b1853a65b97367094ac43a39b7c4bd5dd891ba3a5-filelists.xml.gz - Cannot download, all mirrors were already tried without success
```

mrc0mmand commented:
/packit build

poettering commented on src/core/unit-serialize.c:
```diff
@@ -172,6 +172,29 @@ static int unit_deserialize_job(Unit *u, FILE *f) {
+                bool _deserialize_matched = streq(l, key);              \
+                if (_deserialize_matched) {                             \
+                        if (parse_func(v, &target) < 0)                 \
+                                log_unit_debug(u,  "Failed to parse \"%s=%s\", ignoring", l, v); \
```
double whitespace...

i'd really pass the return error here btw, even if we don't show it in the text message the journal wil pick it up in structured form, and that's still interesting. if we have an error we should pass it to the log subsystem, regardless if we then reference it in the text string with `%m` or not.

poettering commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -1685,6 +1685,8 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
       readonly b IgnoreOnIsolate = ...;
       @org.freedesktop.DBus.Property.EmitsChangedSignal("const")
       readonly b NeedDaemonReload = ...;
+      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
```
we already have a UnitLoadState and a UnitActiveState and and a UnitFileState, it's quite confusing. But these three are actual states as in state machine. But this new StateFlags field isn't realllllllyyy a state in that sense, It's more a marker a flag, but i wouldn't associate the word "state" with it.

Naming things is always the hardest part...

Maybe: "MarkerFlags", "NeedFlags", "IndicatorFlags" or so?

poettering commented on src/core/dbus-unit.c:
```diff
@@ -1839,6 +1839,26 @@ static int bus_unit_set_live_property(
+        /* Two boolean settings that only apply to active units. We don't actually write those to /run, this
+         * state is managed internally. */
+
+        if (STR_IN_SET(name, "NeedsRestart", "NeedsReload")) {
```
hmm, this is quite asymmetric now... So far we tried hard to keep the real bus properties and those passed to SetUnitProperties() as much in sync as we can...

Either this should become also a "StateFlags" (or whatever better name one might find), or StateFlags should become two bools again. it's otherwise too confusing that "systemctl show -p NeedsRestart" fails, but "systemctl set-property NeedsRestart=" works, and vice versa for StateFlags.

This is user visible API after alll

poettering replied:
i'd probably still make this an enum

poettering commented on src/core/dbus-unit.c:
```diff
@@ -1839,6 +1839,26 @@ static int bus_unit_set_live_property(
+         * state is managed internally. */
+
+        if (STR_IN_SET(name, "NeedsRestart", "NeedsReload")) {
+                bool b;
```
sd_bus_message_read() reads an `int`, not a `bool`, this is a bit of a memory corruption...

poettering commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -1171,6 +1174,11 @@ node /org/freedesktop/systemd1 {
       the "Try" flavor is used in which case a service that isn't running is not affected by the restart. The
       "ReloadOrRestart" flavors attempt a reload if the unit supports it and use a restart otherwise.</para>
 
+      <para><function>StartAutomaticJobs()</function> creates reload/restart jobs for units which have been
```
hmm, thing is though, these aren't automatic at all, they require manual triggering after all ;-)

hmm, we also usually start units, not jobs. The jobs are what starting of a unit results in...

also, these don't actually start anything, they reload and restart — only if already running

maybe call this EnqueueMarkedJobs() or so?

poettering commented:
so looks all good to me mostly. but the naming needs some more love. 

maybe we should talk about "staging" in this context, like in the git staging concept.

ie. one bool "StageReload", another "StageRestart" (maybe internally stored in a field stage_flags, but I'd probably not expose this as a flags field on the bus), and then the method call could be called EnqueueStagedJobs.

I think this might be better than the "Needs" moniker, given that this is quite different from the existing "needs" moniker, which is "passive", i.e. not controllable by the user, but these two new bools (or single enum?) are.

keszybz reviewed (commented):
Thank you for the review. I changed the naming around quite a bit, and I'm quite happy with the result.

- I fixed a bug with upgrading of reload to restart when a unit does not support reload. The method to enqueue the jobs (called `method_enqueue_marked_jobs` now), was changed a bit.
- The errnos are logged in deserialization now again. I rebased the whole PR after 751db3b4cd, so the errnos make sense.

The flags fields is now called markers and is exposed as Markers "as" dbus prop. It can be set using `systemctl set-property Markers='+foo -foo'. The two bits are exposes as "needs-restart" and "needs-reload".

The command to restart them is called `systemctl reload-or-restart --marked`.

poettering commented on src/basic/unit-def.c:
```diff
@@ -117,6 +117,13 @@ static const char* const freezer_state_table[_FREEZER_STATE_MAX] = {
+static const char* const unit_marker_table[_UNIT_MARKER_MAX] = {
+        [UNIT_MARKER_NEEDS_RELOAD]  = "needs-reload",
+        [UNIT_MARKER_NEEDS_RESTART] = "needs-restart",
+};
```
hmm, this is a weird table, as it's about a bunch of flags, and thus sparse. this might be ok'ish as long as we only have two flags, but will explode to be a huge sparse table once we add more

poettering replied:
at the very least add a comment here that when somebody adds more here this must be reworked. i.e. with every bit added the table doubles in size. exponential growth sucks...

poettering replied:
or maybe we can index the table here via `fsl(UNIT_MARKER_…)`? But not sure we can express that easily in a constant macro

keszybz replied:
Yeah, I should add a comment. I considered this when writing this, but right now we only waste one entry (i.e. 8 bytes), and it just didn't seems important enough. Obviously with many entries this would be bad, but we might not go above a few flags ever.

poettering commented on src/basic/unit-def.c:
```diff
@@ -117,6 +117,13 @@ static const char* const freezer_state_table[_FREEZER_STATE_MAX] = {
+        [UNIT_MARKER_NEEDS_RESTART] = "needs-restart",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(unit_marker, UnitMarker);
```
it's also a bit weird that unit_marker_to_string() for flag combinations returns 0

poettering commented on src/basic/unit-def.h:
```diff
@@ -58,6 +58,13 @@ typedef enum FreezerState {
 
+typedef enum UnitMarker {
+        UNIT_MARKER_NEEDS_RELOAD  = 1 << 0,
+        UNIT_MARKER_NEEDS_RESTART = 1 << 1,
```
I'd probably define this as regular enum, and then always do (1 << UNIT_MARKER_NEEDS_RELOAD) when we need it as flag, and store the flags in an uint64_t or so

keszybz replied:
OK. I put the flags in unsigned and added a `static_assert()` that this is enough.

poettering commented on src/core/dbus-unit.c:
```diff
@@ -1867,6 +1867,56 @@ static int bus_unit_set_live_property(
+                UnitMarker settings = 0, mask = 0;
+                bool some_plus_minus = false, some_absolute = false;
+
+                r = sd_bus_message_read_strv(message, &s);
```
doesn't matter, but instead of reading the whole thing into an strv and then iterating through it and freeing it again i'd probably just iterate through it one by one right. the code is not that much longer, but should be more efficeint:

```c
r = sd_bus_message_enter(message, 'a', "s");
if (r < 0)
        return r;

for (;;) {
        const char *word;
        r = sd_bus_message_read(message, "s", &word);
        if (r < 0) 
                return r;
        …
}
r = sd_bus_message_exit_container(m);
if (r < 0)
        return r;
```

poettering reviewed (commented):
i don't like the string array indexed by a flags enum I must say, see comments, but i'd be fine if it at least carries a big comment that this needs reworking if any further bit is added, due to exponential array growth in that case.

maybe the string table macros should catch cases like this btw, which might accidentally happen and complain if a string table has more than 100 entries or so. we had a similar case before i think, iirc.

i like the new naming of things!

(we might even think of adding a high level verb one day: "systemctl mark" or so)

poettering commented:
thinking about this, we might want to add more markers eventually, i.e. maybe a "please-clean" (which will do "systemctl clean" right after shutting down a service), "please-reset-failed" (you get it), "please-mask".

or maybe even stuff like "important", which we'd just highlight specially

keszybz commented:
Updated: `unit_marker_table` is now non-sparse. Unit.markers is of type unsigned. A static_assert was added that the defined markers fit.

I also reworked the code to avoid the temporary strv as suggested. It is much nicer.

keszybz commented on src/basic/unit-def.c:
```diff
@@ -117,6 +117,13 @@ static const char* const freezer_state_table[_FREEZER_STATE_MAX] = {
 DEFINE_STRING_TABLE_LOOKUP(freezer_state, FreezerState);
 
+static const char* const unit_marker_table[_UNIT_MARKER_MAX] = {
+        [UNIT_MARKER_NEEDS_RELOAD]  = "needs-reload",
```
I think "needs-*" is better: we're stating a fact. And we ask for an actual reload later on (with the `systemctl reload-or-restart --marked` call.

keszybz commented:
bionic-amd64, bionic-ppc64el failed in TEST-30-ONLCOCKCHANGE.
packit/eln: https://pagure.io/releng/issue/10007.

yuwata reviewed (commented):
One nit.

keszybz commented on src/core/unit-serialize.c:
```diff
@@ -28,6 +28,45 @@ static int serialize_cgroup_mask(FILE *f, const char *key, CGroupMask mask) {
+                        continue;
+                }
+
+                u->markers |= 1 << m;
```
Updated.

keszybz commented:
Anything left to change here?

yuwata commented:
LGTM. But I'd like to leave to push the merge button to @poettering.

poettering commented on src/core/dbus-unit.c:
```diff
@@ -323,6 +323,40 @@ static int property_get_load_error(
+                return r;
+
+        /* Make sure out values fit in the bitfield. */
+        static_assert(_UNIT_MARKER_MAX <= sizeof(((Unit){}).markers) * 8,
```
assert_cc() for compat with pre-C11 compilers

keszybz replied:
Pfff, yeah, I somehow forgot we use assert_cc() for that. Fixed now.

poettering commented on src/core/unit-serialize.c:
```diff
@@ -28,6 +28,45 @@ static int serialize_cgroup_mask(FILE *f, const char *key, CGroupMask mask) {
 }
 
+/* Make sure out values fit in the bitfield. */
+static_assert(_UNIT_MARKER_MAX <= sizeof(((Unit){}).markers) * 8, "Too many UNIT_MARKER_FLAGS defined");
```
ditto

poettering commented:
looks good, just one minor nitpick aroudn assert_cc()/static_assert() left

keszybz commented:
Ooops, it's fuzz-systemctl-parse-argv finding issues. I don't think it's related to the changes in this PR though.

poettering commented:
PR lgtm now.

bluca commented:
FYI: opened a ticket against the Debian tool equivalent of the RPM scriptlets, to see if that too can use this new interface: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=982907

keszybz commented:
So... I think this PR is a good start, but it's certainly not the whole story. We need to do daemon-reexec for the user managers, and also restart units in them. The best idea I have so far is to do the equivalent of
```sh
for u in $(systemctl list-units 'user@*.service'); do
     uid=...
     systemctl -M uid@.host --user set-property %unit Markers=+needs-restart
done
...
for u in $(systemctl list-units 'user@*.service'); do
     uid=...
     systemctl -M uid@.host --user reload-or-restart --marked
done
```
This wouldn't be so bad, except that it calls into unprivileged code from privileged code.
Maybe we can do that, since there's a time limit, not sure.

Better approaches welcome :)

poettering commented:
> This wouldn't be so bad, except that it calls into unprivileged code from privileged code.
> Maybe we can do that, since there's a time limit, not sure.

Yuck. That's terrible... Synchronously calling from privileged code into unprivileged code is terrible...

What might do: introduce a bunch of bus signals one the system bus:

1. InitiateUserDaemonReload()
2. InitiateUserDaemonReexec()
3. InitiateUserMarkUnit(sas)

or so. then make all --user instances install matches for these. And when they get it they check the originator uid: if root it will do what the signal name suggests, otherwise ignore.

Then from rpm scriptlets issue the right `busctl emit` line.

What's key here is that it is a pub/sub scheme, fully async and best-effort: only who's listening will get these notifications, and the caller won't wait for the listeners to handle it.

benjarobin commented on src/basic/unit-def.h:
```diff
@@ -58,6 +58,13 @@ typedef enum FreezerState {
+        UNIT_MARKER_NEEDS_RELOAD,
+        UNIT_MARKER_NEEDS_RESTART,
+        _UNIT_MARKER_MAX,
+        _UNIT_MARKER_INVALID = -1
```
@poettering @keszybz Isn't the new convention to use `-EINVAL` instead of `-1`?

The following elements may need to be analyzed:
```
git grep -P "\b_.*_INVALID *=" | grep -v EINVAL
```
